DigitalLearningSolutions.Data/Services/ActivityService.cs

```diff
@@ -118,12 +124,60 @@ public byte[] GetActivityDataFileForCentre(int centreId, ActivityFilterData filt
         {
             using var workbook = new XLWorkbook();
 
-            var activityData = GetFilteredActivity(centreId, filterData).Select(
-                p => new { Period = p.DateInformation.Date, p.Registrations, p.Completions, p.Evaluations }
-            );
+            var activityData = GetFilteredActivity(centreId, filterData).ToList();
+
+            IEnumerable<WorkbookRow> workbookData;
+
+            if (activityData.Count() <= 1)
+            {
+                workbookData = activityData.Select(
+                    p => new WorkbookRow(
+                        DateInformation.GetDateRangeLabel(
+                            DateHelper.StandardDateFormat,
+                            filterData.StartDate,
+                            filterData.EndDate ?? DateTime.Now
+                        ),
+                        p.Registrations,
+                        p.Completions,
+                        p.Evaluations
+                    )
+                );
+            }
+            else
+            {
+                var first = new[] { activityData.First() }.Select(
+                    p => new WorkbookRow(
+                        p.DateInformation.GetDateRangeLabel(DateHelper.StandardDateFormat, filterData.StartDate, true),
+                        p.Registrations,
+                        p.Completions,
+                        p.Evaluations
+                    )
+                );
+                var last = new[] { activityData.Last() }.Select(
+                    p => new WorkbookRow(
+                        p.DateInformation.GetDateRangeLabel(
+                            DateHelper.StandardDateFormat,
+                            filterData.EndDate ?? DateTime.Now,
+                            true
+                        ),
+                        p.Registrations,
+                        p.Completions,
+                        p.Evaluations
+                    )
+                );

```

Comment:
Same as the view model. Why do we need to use new lists of single items for these rather than just having them be converted directly?

If we're having to do all this messing around with first and last and display strings, would it perhaps make sense to pull all this logic into the GetFilteredActivity method, so we only have this logic in one place.

Reply:
I think moving all the display logic into the GetFilteredActivity method would overload that method, and be an illogical place for it compared to the view model which is the main use case. This logic is only in the service at all because it needs to be done before the file is prepared, which happens in the data project.

It's a little ugly and involves some duplication, but I don't think a completely neat version of this logic is possible, or at least easily achievable at this point.